avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java

```diff
@@ -47,8 +50,33 @@ public abstract class AbstractKafkaAvroDeserializer extends AbstractKafkaSchemaS
   private final DecoderFactory decoderFactory = DecoderFactory.get();
   protected boolean useSpecificAvroReader = false;
   private final Map<String, Schema> readerSchemaCache = new ConcurrentHashMap<>();
-  private final Map<Schema, Map<Schema, DatumReader<?>>> datumReaderCache =
-      new MapMaker().weakKeys().makeMap();  // use identity (==) comparison for keys
+  private final LoadingCache<IdentityPair<Schema, Schema>, DatumReader<?>> datumReaderCache;
+
+  public AbstractKafkaAvroDeserializer() {
+    CacheLoader<IdentityPair<Schema, Schema>, DatumReader<?>> cacheLoader =
+        new CacheLoader<IdentityPair<Schema, Schema>, DatumReader<?>>() {
+          @Override
+          public DatumReader<?> load(IdentityPair<Schema, Schema> key) {
+            Schema writerSchema = key.getKey();
+            Schema readerSchema = key.getValue();
+            Schema finalReaderSchema = getReaderSchema(writerSchema, readerSchema);
+            boolean writerSchemaIsPrimitive =
+                AvroSchemaUtils.getPrimitiveSchemas().containsValue(writerSchema);
+            if (writerSchemaIsPrimitive) {
+              return new GenericDatumReader<>(writerSchema, finalReaderSchema);
+            } else if (useSchemaReflection) {
+              return new ReflectDatumReader<>(writerSchema, finalReaderSchema);
+            } else if (useSpecificAvroReader) {
+              return new SpecificDatumReader<>(writerSchema, finalReaderSchema);
+            } else {
+              return new GenericDatumReader<>(writerSchema, finalReaderSchema);
+            }
+          }
+        };
+    datumReaderCache = CacheBuilder.newBuilder()
+        .maximumSize(DEFAULT_CACHE_CAPACITY)
```

Comment:
I'd assume it needs to be configured, in our scenario we have >1000 schemas.

Reply:
The cache is only an optimization that was only added in 2020.  If you don't hit the cache, the effect should be negligible.  The reason we moved to use identity (=) comparisons is that some customers reported that the equals/hashCode calls actually hurt performance for very large schemas.

Reply:
Sorry I'm basing this on wrong assumption, I'm not super familiar with the code base: If we need to talk to the schema registry for large percentage of the Avro messages we publish, it will be an availability and latency concern for us, and we will be finding an alternative solution.